doc/source/tune/api_docs/trainable.rst

```diff
@@ -1,32 +1,178 @@
+            result_dict = {"diff": abs(self.config['guess'] - self.password)}
+            return result_dict
+
+        def _stop(self):
+            # perform any cleanup necessary.
+            pass
+
+    analysis = tune.run(
+        Guesser,
+        stop={
+            "training_iteration": 1,
+        },
+        num_samples=10,
+        config={
+            "guess": tune.randint(1, 10000)
+        })
+
+    print('best config: ', analysis.get_best_config(metric="diff", mode="min"))
+
+As a subclass of ``tune.Trainable``, Tune will create ``Guesser`` object on a separate process (using the Ray Actor API).
+
+  1. ``_setup`` function is invoked once when training starts.
+  2. ``_train`` is invoked **multiple times**. Each time, the Guesser object executes one logical iteration of training in the tuning process, which may include one or more iterations of actual training.
+  3. ``_stop`` is invoked when training is finished.
+
+.. tip:: As a rule of thumb, the execution time of ``_train`` should be large enough to avoid overheads (i.e. more than a few seconds), but short enough to report progress periodically (i.e. at most a few minutes).
+
+In this example, we only implemented ``_setup`` and ``_train`` methods for simplification. Next, we'll implement ``_save``, and ``_restore`` for checkpoint and fault tolerance.
+
+Save and Restore
+~~~~~~~~~~~~~~~~
+
+Many Tune features rely on ``_save``, and ``_restore``, including the usage of certain Trial Schedulers, fault tolerance, and checkpointing.
+
+.. code-block:: python
+
+    class MyTrainableClass(Trainable):
+        def _save(self, tmp_checkpoint_dir):
+            checkpoint_path = os.path.join(tmp_checkpoint_dir, "model.pth")
+            torch.save(self.model.state_dict(), checkpoint_path)
+            return tmp_checkpoint_dir
+
+        def _restore(self, tmp_checkpoint_dir):
+            checkpoint_path = os.path.join(tmp_checkpoint_dir, "model.pth")
+            self.model.load_state_dict(torch.load(checkpoint_path))
+
+Checkpoints will be saved by training iteration to ``local_dir/exp_name/trial_name/checkpoint_<iter>``. You can restore a single trial checkpoint by using ``tune.run(restore=<checkpoint_dir>)``.
+
+Tune also generates temporary checkpoints for pausing and switching between trials. For this purpose, it is important not to depend on absolute paths in the implementation of ``save``.
+
+Use ``validate_save_restore`` to catch ``_save``/``_restore`` errors before execution.
+
+.. code-block:: python
+
+    from ray.tune.utils import validate_save_restore
+
+    # both of these should return
+    validate_save_restore(MyTrainableClass)
+    validate_save_restore(MyTrainableClass, use_object_store=True)
+
+Advanced: Reusing Actors
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Your Trainable can often take a long time to start. To avoid this, you can use ``tune.run(reuse_actors=True)`` to reuse the same Trainable Python process and object for multiple hyperparameters.
```

Comment:
```suggestion
Your Trainable can often take a long time to start. To avoid this, you can do ``tune.run(reuse_actors=True)`` to reuse the same Trainable Python process and object for multiple hyperparameters.
```